README.txt
/* $Id: README.txt,v 1.8 2009-10-18 12:45:35 snpower Exp $ */

Developer Module that assists with code review and version upgrade that
supports a plug-in extensible hook system so contributed modules can
define additional review standards.

Built-in support for:
 - Drupal Coding Standards - http://drupal.org/node/318
 - Handle text in a secure fashion - http://drupal.org/node/28984
 - Converting 4.6.x modules to 4.7.x - http://drupal.org/node/22218
 - Converting 4.7.x modules to 5.x - http://drupal.org/node/64279

Installation
------------

Copy coder.module to your module directory and then enable on the admin
modules page.  Enable the modules that admin/config/development/coder/settings
works on, then view the coder results page at coder.

Author
------
Doug Green
douggreen@douggreenconsulting.com


CHANGELOG.txt
// $Id: CHANGELOG.txt,v 1.1 2009-10-18 12:45:35 snpower Exp $

Coder 7.x-3.x, xxxx-xx-xx
--------------------------
#607592 (stella) :
 - changed menu paths


scripts/coder_format/tests/CoderTestCase.php
<?php
// $Id: CoderTestCase.php,v 1.2 2008-02-24 19:40:50 sun Exp $

require_once drupal_get_path('module', 'coder') .'/scripts/coder_format/coder_format.inc';

class CoderTestCase extends DrupalTestCase {
  function assertFormat($input, $expect) {
    $result = coder_format_string_all($input);
    $this->assertIdentical($result, $input);
  }
}



scripts/coder_format/tests/CoderTestFile.php
<?php
// $Id: CoderTestFile.php,v 1.2 2008-02-24 19:40:50 sun Exp $


/**
 * Represents coder test file for full coder_format_string_all() tests.
 */
class CoderTestFile extends SimpleExpectation {
  /* Filename of test */
  var $filename;

  /* Test name */
  var $test;

  /* PHP to be parsed */
  var $input = array();

  /* Expected output */
  var $expect = array();

  /* Actual result */
  var $actual = array();

  /* Whether or not <?php and CVS Id should be added */
  var $full = 0;
  
  /* Whether or not a specific test should be the only one tested */
  var $only = array();

  /**
   * Loads this class from a file.
   *
   * @param string $filename
   *   A filename to load.
   */
  function load($filename) {
    $this->filename = $filename;
    $fh             = fopen($filename, 'r');
    $state          = '';
    $unit           = 0;
    
    while (($line = fgets($fh)) !== false) {
      // Normalize newlines.
      $line = rtrim($line, "\n\r");
      // Detect INPUT and EXPECT sections.
      if (substr($line, 0, 2) == '--') {
        $state = trim($line, ' -');
        
        // If a new INPUT section begins, start a new unit.
        if ($state == 'INPUT') {
          // If previous section has been marked with the keyword 'ONLY', break
          // immediately to process only the marked section.
          if ($this->only[$unit]) {
            break;
          }
          $unit++;
        }
        continue;
      }
      // Process other keywords only outside of INPUT and EXPECT sections.
      if (!$state) {
        list($keyword, $line) = explode(': ', $line, 2);
      }
      // Assign previous keyword, if there is no new one.
      else {
        $keyword = $state;
      }
      switch ($keyword) {
        case 'TEST':
          $this->test = $line;
          break;

        case 'FULL':
          $this->full = (bool)$line;
          break;

        case 'INPUT':
          $this->input[$unit] .= $line ."\n";
          break;

        case 'EXPECT':
          $this->expect[$unit] .= $line ."\n";
          break;
        
        case 'ONLY':
          $this->only[$unit] = TRUE;
          break;
      }
    }
    fclose($fh);
    foreach (range(1, $unit) as $unit) {
      // If no EXPECTed code was defined, INPUT is expected.
      if (!isset($this->expect[$unit])) {
        $this->expect[$unit] = $this->input[$unit];
      }
      // If FULL was *not* defined, add a PHP header to contents.
      if (!$this->full) {
        $prepend             = "<?php\n// $". "Id$\n\n";
        $this->input[$unit]  = $prepend . rtrim($this->input[$unit], "\n") ."\n\n";
        $this->expect[$unit] = $prepend . rtrim($this->expect[$unit], "\n") ."\n\n";
      }
    }
    if (!empty($this->only[$unit])) {
      $this->input = array($this->input[$unit]);
      $this->expect = array($this->expect[$unit]);
    }
  }

  /**
   * Implements SimpleExpectation::test().
   *
   * @param $filename Filename of test file to test.
   */
  function test($filename = false) {
    if ($filename) {
      $this->load($filename);
    }
    
    // Perform test.
    // Test passes until proven invalid.
    $valid = TRUE;
    foreach ($this->input as $unit => $content) {
      // Parse input and store results.
      $this->actual[$unit] = coder_format_string_all($this->input[$unit]);
      
      // Let this test fail, if a unit fails.
      if ($this->expect[$unit] !== $this->actual[$unit]) {
        $valid = FALSE;
      }
    }
    
    return $valid;
  }

  /**
   * Implements SimpleExpectation::testMessage().
   */
  function testMessage() {
    $message = $this->test .' test in '. htmlspecialchars(basename($this->filename));
    return $message;
  }

  /**
   * Renders the test with an HTML diff table.
   */
  function render() {
    drupal_add_css(drupal_get_path('module', 'coder') .'/scripts/coder_format/tests/coder-diff.css', 'module', 'all', false);
    
    foreach ($this->input as $unit => $content) {
      // Do not output passed units.
      if ($this->expect[$unit] === $this->actual[$unit]) {
        continue;
      }
      
      $diff     = new Text_Diff('auto', array(explode("\n", $this->expect[$unit]), explode("\n", $this->actual[$unit])));
      $renderer = new Text_Diff_Renderer_parallel($this->test .' test in '. htmlspecialchars(basename($this->filename)));
      
      $message .= $renderer->render($diff);
    }
    
    return $message;
  }
}

/**
 * Parallel diff renderer for HTML tables with original text on left,
 * new text on right, and changed text highlighted with appropriate classes.
 */
class Text_Diff_Renderer_parallel extends Text_Diff_Renderer {
  /* String header for left column */
  var $original = 'Expected';

  /* String header for right column */
  var $final = 'Actual';
  
  // These are big to ensure entire string is output.
  var $_leading_context_lines  = 10000;
  var $_trailing_context_lines = 10000;
  var $title;
  
  function Text_Diff_Renderer_parallel($title) {
    $this->title = $title;
  }

  function _blockHeader() {}

  function _startDiff() {
    return '<table class="diff"><thead><tr><th colspan="2">'. $this->title .'</th></tr><tr><th>'. $this->original .'</th><th>'. $this->final .'</th></tr></thead><tbody>';
  }

  function _endDiff() {
    return '</tbody></table>';
  }

  function _context($lines) {
    return '<tr><td><pre>'. $this->_renderLines($lines) .'</pre></td>
          <td><pre>'. $this->_renderLines($lines) .'</pre></td></tr>';
  }

  function _added($lines) {
    return '<tr><td>&nbsp;</td><td class="added"><pre>'. $this->_renderLines($lines) .'</pre></td></tr>';
  }

  function _deleted($lines) {
    return '<tr><td class="deleted"><pre>'. $this->_renderLines($lines) .'</pre></td><td>&nbsp;</td></tr>';
  }

  function _changed($orig, $final) {
    return '<tr class="changed"><td><pre>'. $this->_renderLines($orig) .'</pre></td>
        <td><pre>'. $this->_renderLines($final) .'</pre></td></tr>';
  }
  
  function _renderLines($lines) {
    return str_replace("\n", "<strong>&para;</strong>\n", htmlspecialchars(implode("\n", $lines)."\n"));
  }
  
}



coder.js
// $Id: coder.js,v 1.4 2009-03-04 15:57:23 snpower Exp $
(function ($) {
  jQuery.fn.extend({
    check : function() { return this.each(function() { this.checked = true; }); },
    uncheck : function() { return this.each(function() { this.checked = false; }); }
  });

  Drupal.behaviors.coderFormHandler = {
    attach: function (context) {
      $("input:checkbox").click(
        function() {
          core = this.form.elements.namedItem("edit-coder-core");
          active = this.form.elements.namedItem("edit-coder-active-modules");
          if (this == core || this == active) {
            modules = "input[@id^=edit-coder-modules-]";
            themes = "input[@id^=edit-coder-themes-]";
            if (core.checked || active.checked) {
              $(modules).uncheck();
              $(themes).uncheck();
              if (core.checked) {
                modules += '.coder-core';
                themes += '.coder-core';
              }
              if (active.checked) {
                modules += '.coder-active';
                themes += '.coder-active';
              }
              $(modules).check();
              $(themes).check();
            }
            else {
              if (this == active) {
                modules += ".coder-active";
                themes += ".coder-active";
              }
              else {
                modules += ".coder-core";
                themes += ".coder-core";
              }
              $(modules).uncheck();
              $(themes).uncheck();
            }
          }
          else if (this.id.substr(0, 19) == "edit-coder-modules-" || this.id.substr(0, 18) == "edit-coder-themes-") {
            core.checked = false;
            active.checked = false;
          }
          return true;
        }
      );
      $("img.coder-more").click(
        function() {
          $('.coder-description', this.parentNode).slideToggle();
        }
      );
    }
  };
})(jQuery);


coder_review/coder_review.js
// $Id: coder_review.js,v 1.2 2009-08-08 14:38:56 snpower Exp $
(function ($) {
  jQuery.fn.extend({
    check : function() { return this.each(function() { this.checked = true; }); },
    uncheck : function() { return this.each(function() { this.checked = false; }); }
  });

  Drupal.behaviors.coderReviewFormHandler = {
    attach: function (context) {
      $("input:checkbox").click(
        function() {
          core = this.form.elements.namedItem("edit-coder-core");
          active = this.form.elements.namedItem("edit-coder-active-modules");
          if (this == core || this == active) {
            modules = "input[id^=edit-coder-modules-]";
            themes = "input[id^=edit-coder-themes-]";
            if (core.checked || active.checked) {
              $(modules).uncheck();
              $(themes).uncheck();
              if (core.checked) {
                modules += '.coder-core';
                themes += '.coder-core';
              }
              if (active.checked) {
                modules += '.coder-active';
                themes += '.coder-active';
              }
              $(modules).check();
              $(themes).check();
            }
            else {
              if (this == active) {
                modules += ".coder-active";
                themes += ".coder-active";
              }
              else {
                modules += ".coder-core";
                themes += ".coder-core";
              }
              $(modules).uncheck();
              $(themes).uncheck();
            }
          }
          else if (this.id.substr(0, 19) == "edit-coder-modules-" || this.id.substr(0, 18) == "edit-coder-themes-") {
            core.checked = false;
            active.checked = false;
          }
          return true;
        }
      );
      $("img.coder-more").click(
        function() {
          $('.coder-description', this.parentNode).slideToggle();
        }
      );
    }
  };
})(jQuery);


coder_upgrade/coder_upgrade.api.php
<?php
// $Id: coder_upgrade.api.php,v 1.4 2009-08-17 16:04:46 solotandem Exp $

/**
 * @file
 * Hooks provided by the Coder Upgrade module.
 *
 * Copyright 2008-9 by Jim Berry ("solotandem", http://drupal.org/user/240748)
 */

/**
 * @addtogroup hooks
 * @{
 */

/**
 * Declare upgrade sets for an API (or set of APIs).
 *
 * This hook allows contributed modules to declare upgrade sets for an API
 * supplied by their module, another contributed module, or a core module. Each
 * upgrade set may optionally include a list of conversion routines.
 *
 * For example, if your module is called "your_module_name," then declare an
 * upgrade set as:
 * @code
 *   function your_module_name_upgrades($include_routines) {
 *     $routines = $include_routines ? your_module_name_conversions_list() : array();
 *     $upgrade = array(
 *       'title' => t('Your module API changes from 6.x to 7.x'),
 *       'link' => 'http://...',
 *       'routines' => $routines,
 *     );
 *     return array('your_module_name' => $upgrade);
 *   }
 * @endcode
 *
 * The naming convention for a conversion routine is:
 * @code
 *   your_module_name_convert_your_routine_name();
 * @endcode
 *
 * The array returned by your_module_name_conversions_list() should be a list
 * of the suffixes of your conversion routines. The suffix is the portion of
 * the routine name following "_convert_" above.
 *
 * Group changes by processing order (beginning, middle, or end). The beginning
 * and end conversion routines are applied only at the module or directory
 * level and the routine will need to determine what to do and how (regex or
 * parser) to do it.
 *
 * The middle changes are applied at the file level and the list entry must
 * specify the handler (regex or parser). The regex changes are applied before
 * the parser changes.
 * @code
 *   function your_module_name_conversions_list() {
 *     $list = array(
 *       'beginning' => array(),
 *       'middle' => array(),
 *       'end' => array(),
 *     );
 *
 *     $list['middle'] = array(
 *       array(
 *         'name' => 'your_routine_name_regex',
 *         'type' => 'regex',
 *       ),
 *       array(
 *         'name' => 'your_routine_name_parser',
 *         'type' => 'parser',
 *       ),
 *     );
 *
 *     return $list;
 *   }
 * @endcode
 *
 * The $file parameter passed to your regex conversion routine is a reference
 * to the text of the code file to be upgraded. Be sure to include an ampersand
 * before $file so that your changes will be incorporated into the output code
 * file.
 * @code
 *   function your_module_name_convert_your_routine_name_regex(&$file) {
 *     // Do something to $file.
 *   }
 * @endcode
 *
 * The $reader parameter passed to your parser conversion routine is a
 * reference to the object containing the grammar statements of the file to
 * convert. Be sure to include an ampersand before $reader so that your changes
 * will be incorporated into the output code file.
 * @code
 *   function your_module_name_convert_your_routine_name_parser(&$reader) {
 *     // Do something with $reader.
 *   }
 * @endcode
 *
 * To use the sample functions included in this api file:
 * - Copy and paste the sample functions to a file in your module.
 * - Replace "your_module_name" with your actual module name.
 * - Complete the conversions_list() routine.
 * - Duplicate the sample conversion routine for each entry in your list,
 *   replacing "your_routine_name" with the appropriate value and changing the
 *   comment block to describe the upgrade applied by the routine.
 *
 * @param boolean $include_routines
 *   Indicates whether to include the list of conversion routines. This list
 *   is only needed when the conversions are to be applied (typically on form
 *   submission). This parameter equals FALSE when this hook is invoked by
 *   Coder Upgrade while building the form, and TRUE when applying the
 *   conversions.
 */
function hook_upgrades($include_routines)  {
  $routines = $include_routines ? your_module_name_conversions_list() : array();
  $upgrade = array(
    'title' => t('Your module API changes from 6.x to 7.x'),
    'link' => 'http://...',
    'routines' => $routines,
  );
  return array('your_module_name' => $upgrade);
}

/**
 * @} End of "addtogroup hooks".
 */

/**
 * Sample functions.
 */

/**
 * Implement hook_upgrades().
 */
function your_module_name_upgrades($include_routines)  {
  $routines = $include_routines ? your_module_name_conversions_list() : array();
  $upgrade = array(
    'title' => t('Your module API changes from 6.x to 7.x'),
    'link' => 'http://...',
    'routines' => $routines,
  );
  return array('your_module_name' => $upgrade);
}

/**
 * Return a list of conversion routine suffixes.
 *
 * The complete function name that Coder Upgrade searches for is:
 *   <your_module_name> . '_convert_' . <your_routine_name>
 *
 * The entries in the array returned from this routine should only contain
 * the <your_routine_name> portion.
 *
 * @return array
 *   Array of conversion routine suffixes.
 */
function your_module_name_conversions_list() {
  $list = array(
    'beginning' => array(),
    'middle' => array(),
    'end' => array(),
  );

  $list['middle'] = array(
    array(
      'name' => 'your_routine_name_regex',
      'type' => 'regex',
    ),
    array(
      'name' => 'your_routine_name_parser',
      'type' => 'parser',
    ),
  );

  return $list;
}

/**
 * Describe the upgrade applied by this routine.
 *
 * @param string $file
 *   The text of the file to convert.
 */
function your_module_name_convert_your_routine_name_regex(&$file) {
  // Do something to $file.
  $hook = 'your_changes'; // Used as the label in the log file.
  $cur = $file;
  $new = $cur;

  $from = array();
  $to = array();

  $from[] = '/(your_module_name)/';
  $to[] = "$1";

  coder_upgrade_do_conversions($from, $to, $new);
  coder_upgrade_save_changes($cur, $new, $file, $hook);
}

/**
 * Describe the upgrade applied by this routine.
 *
 * @param PGPReader $reader
 *   The object containing the grammar statements of the file to convert.
 */
function your_module_name_convert_your_routine_name_parser(&$reader) {
  // Do something with $reader.
}


scripts/coder_format/coder_format.php
<?php
// $Id: coder_format.php,v 1.3 2008-02-24 19:40:50 sun Exp $


/**
 * @file
 * Coder format shell invocation script.
 *
 * @param $filepath
 * 	 One or more files or directories containing the source code to process. If
 *   a directory is given, all files in that directory are processed recursively.
 *   Each file will be replaced with the formatted source code. Each original
 *   file is automatically backup to <filename>.coder.orig.
 * @param --undo
 *   If the optional --undo argument is given, processed files are restored from
 *   their backups. Coder format automatically searches for the latest backup
 *   file (*.coder.orig) and each filename (without extension .coder.orig) is
 *   replaced with its backup file.
 *
 * @usage
 *   php coder_format.php <filepath> [filepath] ...
 *   php coder_format.php <path> [path] ...
 *   php coder_format.php --undo <filepath> ...
 *   php coder_format.php --undo <path> ...
 *
 * @example
 *   php coder_format.php modules/node/node.module
 *   php coder_format.php index.php modules/node/node.module
 *   php coder_format.php /home/drupal
 *   php coder_format.php /home/drupal/includes /home/drupal/modules
 *   php coder_format.php --undo modules/node/node.module
 *   php coder_format.php --undo /home/drupal
 *
 * Important note for Windows users:
 * - Ensure to encapsulate filename arguments with double quotes.
 */

// We are on the command line, so output only real errors.
error_reporting(E_ERROR | E_WARNING | E_PARSE);

require_once realpath(dirname($_SERVER['PHP_SELF'])) .'/coder_format.inc';

if (!empty($_SERVER['argv'])) {
  // Remove self-reference.
  array_shift($_SERVER['argv']);

  // Process command-line arguments.
  $files    = array();
  $undo     = false;

  for ($c = 0, $cc = count($_SERVER['argv']); $c < $cc; ++$c) {
    switch ($_SERVER['argv'][$c]) {
      case '--undo':
        $undo = true;
        break;

      default:
        $files[] = $_SERVER['argv'][$c];
        break;
    }
  }

  foreach ($files as $file) {
    // If argument is a directory, process it recursively.
    if (is_dir($file)) {
      coder_format_recursive($file, $undo);
    }
    else {
      coder_format_file($file, $undo);
    }
  }
}

/**
 * @defgroup coder_format_file_functions
 * @{
 * These functions are copied from Drupal's file.inc. Almost all function calls
 * to other Drupal functions have been removed since they would always return
 * FALSE.
 */

define('FILE_CREATE_DIRECTORY', 1);
define('FILE_EXISTS_RENAME', 0);
define('FILE_EXISTS_REPLACE', 1);
define('FILE_EXISTS_ERROR', 2);

/**
 * Check that the directory exists and is writable. Directories need to
 * have execute permissions to be considered a directory by FTP servers, etc.
 *
 * @param $directory A string containing the name of a directory path.
 * @param $mode A Boolean value to indicate if the directory should be created
 *   if it does not exist or made writable if it is read-only.
 * @param $form_item An optional string containing the name of a form item that
 *   any errors will be attached to. This is useful for settings forms that
 *   require the user to specify a writable directory. If it can't be made to
 *   work, a form error will be set preventing them from saving the settings.
 * @return FALSE when directory not found, or TRUE when directory exists.
 */
function file_check_directory(&$directory, $mode = 0, $form_item = NULL) {
  $directory = rtrim($directory, '/\\');

  // Check if directory exists.
  if (!is_dir($directory)) {
    // coder_format does not alter the filesystem. 23/01/2008 sun
    if (!file_exists($directory)) {
      drupal_set_message(t('The directory %directory does not exist.', array('%directory' => $directory)));
      return FALSE;
    }
  }

  // Check to see if the directory is writable.
  if (!is_writable($directory)) {
    // coder_format does not alter the filesystem. 23/01/2008 sun
    drupal_set_message(t('The directory %directory is not writable', array('%directory' => $directory)));
    return FALSE;
  }

  // coder_format is applied outside of Drupal in most cases. 23/01/2008 sun

  return TRUE;
}

/**
 * Checks path to see if it is a directory, or a dir/file.
 *
 * @param $path A string containing a file path. This will be set to the
 *   directory's path.
 * @return If the directory is not in a Drupal writable directory, FALSE is
 *   returned. Otherwise, the base name of the path is returned.
 */
function file_check_path(&$path) {
  // Check if path is a directory.
  // coder_format uses file_check_path() to check for files only. 26/01/2008 sun

  // Check if path is a possible dir/file.
  $filename = basename($path);
  $path = dirname($path);
  if (file_check_directory($path)) {
    return $filename;
  }

  return FALSE;
}

/**
 * Copies a file to a new location. This is a powerful function that in many ways
 * performs like an advanced version of copy().
 * - Checks if $source and $dest are valid and readable/writable.
 * - Performs a file copy if $source is not equal to $dest.
 * - If file already exists in $dest either the call will error out, replace the
 *   file or rename the file based on the $replace parameter.
 *
 * @param $source A string specifying the file location of the original file.
 *   This parameter will contain the resulting destination filename in case of
 *   success.
 * @param $dest A string containing the directory $source should be copied to.
 *   If this value is omitted, Drupal's 'files' directory will be used.
 * @param $replace Replace behavior when the destination file already exists.
 *   - FILE_EXISTS_REPLACE - Replace the existing file
 *   - FILE_EXISTS_RENAME - Append _{incrementing number} until the filename is unique
 *   - FILE_EXISTS_ERROR - Do nothing and return FALSE.
 * @return True for success, FALSE for failure.
 */
function file_copy(&$source, $dest = 0, $replace = FILE_EXISTS_RENAME) {
  // $dest is almost always outside of Drupal. 23/01/2008 sun
  // $dest = file_create_path($dest);

  $directory = $dest;
  $basename = file_check_path($directory);

  // Make sure we at least have a valid directory.
  if ($basename === FALSE) {
    $source = is_object($source) ? $source->filepath : $source;
    drupal_set_message(t('The selected file %file could not be uploaded, because the destination %directory is not properly configured.', array('%file' => $source, '%directory' => $dest)), 'error');
    return 0;
  }

  // Removed upload handling. coder_format does not deal with uploads. 23/01/2008 sun

  $source = realpath($source);
  if (!file_exists($source)) {
    drupal_set_message(t('The selected file %file could not be copied, because no file by that name exists. Please check that you supplied the correct filename.', array('%file' => $source)), 'error');
    return 0;
  }

  // If the destination file is not specified then use the filename of the source file.
  $basename = $basename ? $basename : basename($source);
  $dest = $directory .'/'. $basename;

  // Make sure source and destination filenames are not the same, makes no sense
  // to copy it if they are. In fact copying the file will most likely result in
  // a 0 byte file. Which is bad. Real bad.
  if ($source != realpath($dest)) {
    if (!$dest = file_destination($dest, $replace)) {
      drupal_set_message(t('The selected file %file could not be copied, because a file by that name already exists in the destination.', array('%file' => $source)), 'error');
      return FALSE;
    }

    if (!@copy($source, $dest)) {
      drupal_set_message(t('The selected file %file could not be copied.', array('%file' => $source)), 'error');
      return 0;
    }

    // Give everyone read access so that FTP'd users or
    // non-webserver users can see/read these files,
    // and give group write permissions so group members
    // can alter files uploaded by the webserver.
    @chmod($dest, 0664);
  }

  // Removed upload handling. coder_format does not deal with uploads. 23/01/2008 sun
  $source = $dest;

  return 1; // Everything went ok.
}

/**
 * Determines the destination path for a file depending on how replacement of
 * existing files should be handled.
 *
 * @param $destination A string specifying the desired path.
 * @param $replace Replace behavior when the destination file already exists.
 *   - FILE_EXISTS_REPLACE - Replace the existing file
 *   - FILE_EXISTS_RENAME - Append _{incrementing number} until the filename is
 *     unique
 *   - FILE_EXISTS_ERROR - Do nothing and return FALSE.
 * @return The destination file path or FALSE if the file already exists and
 *   FILE_EXISTS_ERROR was specified.
 */
function file_destination($destination, $replace) {
  if (file_exists($destination)) {
    switch ($replace) {
      case FILE_EXISTS_RENAME:
        $basename = basename($destination);
        $directory = dirname($destination);
        $destination = file_create_filename($basename, $directory);
        break;

      case FILE_EXISTS_ERROR:
        drupal_set_message(t('The selected file %file could not be copied, because a file by that name already exists in the destination.', array('%file' => $destination)), 'error');
        return FALSE;
    }
  }
  return $destination;
}

/**
 * Moves a file to a new location.
 * - Checks if $source and $dest are valid and readable/writable.
 * - Performs a file move if $source is not equal to $dest.
 * - If file already exists in $dest either the call will error out, replace the
 *   file or rename the file based on the $replace parameter.
 *
 * @param $source A string specifying the file location of the original file.
 *   This parameter will contain the resulting destination filename in case of
 *   success.
 * @param $dest A string containing the directory $source should be copied to.
 *   If this value is omitted, Drupal's 'files' directory will be used.
 * @param $replace Replace behavior when the destination file already exists.
 *   - FILE_EXISTS_REPLACE - Replace the existing file
 *   - FILE_EXISTS_RENAME - Append _{incrementing number} until the filename is unique
 *   - FILE_EXISTS_ERROR - Do nothing and return FALSE.
 * @return True for success, FALSE for failure.
 */
function file_move(&$source, $dest = 0, $replace = FILE_EXISTS_RENAME) {
  $path_original = is_object($source) ? $source->filepath : $source;

  if (file_copy($source, $dest, $replace)) {
    $path_current = is_object($source) ? $source->filepath : $source;

    if ($path_original == $path_current || file_delete($path_original)) {
      return 1;
    }
    drupal_set_message(t('The removal of the original file %file has failed.', array('%file' => $path_original)), 'error');
  }
  return 0;
}

/**
 * Create a full file path from a directory and filename. If a file with the
 * specified name already exists, an alternative will be used.
 *
 * @param $basename string filename
 * @param $directory string directory
 * @return
 */
function file_create_filename($basename, $directory) {
  $dest = $directory .'/'. $basename;

  if (file_exists($dest)) {
    // Destination file already exists, generate an alternative.
    // Always append '.coder.orig' (allows multiple undos). 23/01/2008 sun
    $name = $basename;

    $counter = 0;
    do {
      $dest = $directory .'/'. $name . str_repeat('.coder.orig', $counter++);
    } while (file_exists($dest));
  }

  return $dest;
}

/**
 * Delete a file.
 *
 * @param $path A string containing a file path.
 * @return TRUE for success, FALSE for failure.
 */
function file_delete($path) {
  if (is_file($path)) {
    return unlink($path);
  }
}

/**
 * Finds all files that match a given mask in a given directory.
 * Directories and files beginning with a period are excluded; this
 * prevents hidden files and directories (such as SVN working directories)
 * from being scanned.
 *
 * @param $dir
 *   The base directory for the scan, without trailing slash.
 * @param $mask
 *   The regular expression of the files to find.
 * @param $nomask
 *   An array of files/directories to ignore.
 * @param $callback
 *   The callback function to call for each match.
 * @param $recurse
 *   When TRUE, the directory scan will recurse the entire tree
 *   starting at the provided directory.
 * @param $key
 *   The key to be used for the returned array of files. Possible
 *   values are "filename", for the path starting with $dir,
 *   "basename", for the basename of the file, and "name" for the name
 *   of the file without an extension.
 * @param $min_depth
 *   Minimum depth of directories to return files from.
 * @param $depth
 *   Current depth of recursion. This parameter is only used internally and should not be passed.
 *
 * @return
 *   An associative array (keyed on the provided key) of objects with
 *   "path", "basename", and "name" members corresponding to the
 *   matching files.
 */
function file_scan_directory($dir, $mask, $nomask = array('.', '..', 'CVS'), $callback = 0, $recurse = TRUE, $key = 'filename', $min_depth = 0, $depth = 0) {
  $key = (in_array($key, array('filename', 'basename', 'name')) ? $key : 'filename');
  $files = array();

  if (is_dir($dir) && $handle = opendir($dir)) {
    while ($file = readdir($handle)) {
      if (!in_array($file, $nomask) && $file[0] != '.') {
        if (is_dir("$dir/$file") && $recurse) {
          // Give priority to files in this folder by merging them in after any subdirectory files.
          $files = array_merge(file_scan_directory("$dir/$file", $mask, $nomask, $callback, $recurse, $key, $min_depth, $depth + 1), $files);
        }
        elseif ($depth >= $min_depth && ereg($mask, $file)) {
          // Always use this match over anything already set in $files with the same $$key.
          $filename = "$dir/$file";
          $basename = basename($file);
          $name = substr($basename, 0, strrpos($basename, '.'));
          $files[$$key] = new stdClass();
          $files[$$key]->filename = $filename;
          $files[$$key]->basename = $basename;
          $files[$$key]->name = $name;
          if ($callback) {
            $callback($filename);
          }
        }
      }
    }

    closedir($handle);
  }

  return $files;
}

/**
 * @} End of "defgroup coder_format_file_functions".
 */

/**
 * @defgroup coder_format_stub_functions
 * @{
 */
function t($string, $args = 0) {
  if (!$args) {
    return $string;
  }
  else {
    return strtr($string, $args);
  }
}

function drupal_set_message($message = NULL, $type = 'status') {
  if ($type == 'error') {
    echo str_repeat('-', 80);
    echo "\nERROR: $message\n";
    echo str_repeat('-', 80);
    echo "\n";
  }
  else {
    echo "$message\n";
  }
}

/**
 * @} End of "defgroup coder_format_stub_functions".
 */

